metricflow-semantics/metricflow_semantics/test_helpers/performance_helpers.py

```diff
@@ -0,0 +1,350 @@
+
+    def add_report(self, report: SessionReport) -> None:
+        """Add a report and associate it with the session ID."""
+        self.sessions[report.session_id] = report
+
+    def compare(self, other: SessionReportSet) -> SessionReportSetComparison:
+        """Compare this report set with other."""
+        self_sessions = set(self.sessions.keys())
+        other_sessions = set(other.sessions.keys())
+        all_sessions = self_sessions.union(other_sessions)
+
+        comparison: Dict[str, Optional[SessionReportComparison]] = {}
+        max_pct_change = float("-inf")
+        max_pct_change_session = ""
+        for session in all_sessions:
+            if session not in self.sessions or session not in other.sessions:
+                comparison[session] = None
+            else:
+                comp = self.sessions[session].compare(other.sessions[session])
+                comparison[session] = comp
+                if comp.max_pct_change > max_pct_change:
+                    max_pct_change = comp.max_pct_change
+                    max_pct_change_session = session
+
+        return SessionReportSetComparison(
+            sessions=comparison,
+            max_pct_change=max_pct_change,
+            max_pct_change_session=max_pct_change_session,
+        )
+
+
+class SessionReportSetComparison(FrozenBaseModel):
+    """A comparison between two session report sets.
+
+    If a session ID is not present in A or B, the comparison is None
+    """
+
+    sessions: Dict[str, Optional[SessionReportComparison]]
+
+    max_pct_change: float
+    max_pct_change_session: str
+
+
+class PerformanceTracker:
+    """Track performance metrics across different contexts.
+
+    Don't use this directly. Instead, use the global methods in this method which interact
+    with the _perf singleton.
+    """
+
+    def __init__(self) -> None:
+        """Initialize the tracker."""
+        self._session_id: Optional[str] = None
+        self._call_map: Dict[str, List[Call]] = defaultdict(list)
+
+        self._session_set = SessionReportSet()
+
+    @contextmanager
+    def session(self, session_id: str) -> Iterator[None]:
+        """Create a new measurement session.
+
+        At session start, all state is fresh and it gets cleaned when the session ends.
+        """
+        if self._session_id:
+            raise ValueError("Cannot create nested sessions.")
+
+        self._session_id = session_id
+
+        yield
+
+        report = self.get_session_report()
+        self._session_set.add_report(report)
+
+        self._call_map = defaultdict(list)
+        self._session_id = None
+
+    @contextmanager
+    def measure(self, context_id: str) -> Iterator[PerformanceTracker]:
+        """Measure performance while executing this block."""
+        if not self._session_id:
+            raise ValueError("Cannot measure outside of a session.")
+
+        start_wall = time.time_ns()
+        start_cpu = time.process_time_ns()
+
+        yield self
```

Comment:
High level question - what's the advantage of doing this ourself instead of using the output of `cProfile`?